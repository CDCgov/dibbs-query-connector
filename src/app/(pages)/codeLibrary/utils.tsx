/**
 * Helper method to extract the source system display name from the url string.
 * Similar to DBServive.stripProtocolAndTLDFromSystemUrl, but builds in specific
 * checks for known string outliers
 * @param system the url string to format
 * @returns the display name of the system, stripped of url prefixes
 */
export const formatSystem = (system: string) => {
  const match = system.match(/https?:\/\/([^\.]+)/);
  let result = match?.[1];

  switch (result) {
    case "hl7":
      const hl7Arr = system.split("/");
      result = hl7Arr[hl7Arr.length - 1];
      if (result == "icd-10-cm") {
        result = result.slice(0, result.indexOf("-cm"));
      }
      return result.toLocaleUpperCase();

    case "www":
      const rxArr = system.split("/");
      return rxArr[rxArr?.length - 1].toLocaleUpperCase();

    default:
      return result?.toUpperCase();
  }
};
