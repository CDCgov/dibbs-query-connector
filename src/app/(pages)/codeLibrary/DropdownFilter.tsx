import { Dispatch, RefObject, SetStateAction, useEffect, useRef } from "react";
import { Button, Select } from "@trussworks/react-uswds";
import styles from "./codeLibrary.module.scss";
import {
  DibbsConceptType,
  DibbsValueSet,
} from "@/app/models/entities/valuesets";
import { formatStringToSentenceCase } from "@/app/shared/format-service";
import { formatSystem } from "./utils";

export type FilterCategories = {
  category: DibbsConceptType | undefined;
  codeSystem: string;
  creator: string;
};

type DropdownFilterProps = {
  setFilterSearch: Dispatch<SetStateAction<FilterCategories>>;
  filterSearch: FilterCategories;
  valueSets: DibbsValueSet[];
  setShowFilters: Dispatch<SetStateAction<boolean>>;
  loading: boolean;
  filterCount: number;
};

/**
 * @param root0 props
 * @param root0.filterSearch the filter criteria
 * @param root0.setFilterSearch state function to update filter criteria
 * @param root0.setShowFilters state function to toggle showing/hiding the dropdown filters
 * @param root0.valueSets the value sets to apply the filter(s) to
 * @param root0.loading the loading state of the parent's value set data
 * @param root0.filterCount the number of filters currently applied to the result set
 * @returns  the DropdownFilter component
 */
const DropdownFilter: React.FC<DropdownFilterProps> = ({
  filterSearch,
  setFilterSearch,
  valueSets,
  setShowFilters,
  loading,
  filterCount,
}) => {
  const valueSetCategories: {
    [dibbsConceptType in DibbsConceptType]: dibbsConceptType;
  } = {
    labs: "labs",
    medications: "medications",
    conditions: "conditions",
  };
  const filterShortcut = (type: string) => {
    // TODO: logic to support "created by me" / "created by my team" filtering
    setFilterSearch({ ...filterSearch, creator: type });
  };

  const valueSetCodeSystems = valueSets
    .map((vs) => vs.system)
    .filter((item, index, array) => {
      return array.indexOf(item) === index;
    });

  const valueSetCreators = valueSets
    .map((vs) => vs.author)
    .filter((item, index, array) => {
      return array.indexOf(item) === index;
    });

  const handleOutsideClick = (ref: RefObject<HTMLFormElement>) => {
    useEffect(() => {
      function handleClickOutside(event: MouseEvent | KeyboardEvent) {
        if (
          (event as KeyboardEvent).key == "Escape" ||
          (ref.current && !ref.current.contains(event.target as Node))
        ) {
          return setShowFilters(false);
        }
      }
      document.addEventListener("mousedown", handleClickOutside);
      document.addEventListener("keyup", handleClickOutside);
      return () => {
        document.removeEventListener("mousedown", handleClickOutside);
        document.removeEventListener("keyup", handleClickOutside);
      };
    }, [ref]);
  };

  const dropdownRef = useRef<HTMLFormElement>(null);
  handleOutsideClick(dropdownRef);

  return (
    <form ref={dropdownRef} className={styles.filtersDropdown}>
      <div className={styles.filterOptions}>
        <div className={styles.filterOptions_grouping}>
          <label htmlFor="category">Category</label>
          <Select
            id="category"
            name="category"
            className={styles.filtersSelect}
            onChange={(e) =>
              setFilterSearch({
                ...filterSearch,
                category: e.target.value as DibbsConceptType,
              })
            }
            value={filterSearch.category}
            disabled={!!loading}
          >
            <option value=""></option>
            {Object.keys(valueSetCategories).map((category) => {
              return (
                <option key={category} value={category}>
                  {formatStringToSentenceCase(category)}
                </option>
              );
            })}
          </Select>
        </div>
        <div className={styles.filterOptions_grouping}>
          <label htmlFor="code-system">Code system</label>
          <Select
            id="code-system"
            className={styles.filtersSelect}
            name="code-system"
            onChange={(e) => {
              e.preventDefault();
              return setFilterSearch({
                ...filterSearch,
                codeSystem: e.target.value as DibbsConceptType,
              });
            }}
            value={filterSearch.codeSystem}
            disabled={!!loading}
          >
            {<option value=""></option>}
            {valueSetCodeSystems.map((codeSystem) => {
              return (
                <option key={codeSystem} value={codeSystem}>
                  {formatSystem(codeSystem)}
                </option>
              );
            })}
          </Select>
        </div>
        <div className={styles.filterOptions_grouping}>
          <label htmlFor="creator">Creator</label>
          <Select
            id="creator"
            name="creator"
            className={styles.filtersSelect}
            onChange={(e) =>
              setFilterSearch({
                ...filterSearch,
                creator: e.target.value as DibbsConceptType,
              })
            }
            value={filterSearch.creator}
            disabled={!!loading}
          >
            <option value=""></option>
            {valueSetCreators.map((creator) => {
              return <option key={creator}>{creator}</option>;
            })}
          </Select>
        </div>
      </div>
      <div className={styles.shortcuts}>
        <span className="text-italic">Shortcuts:</span>
        <button
          onClick={(e) => {
            e.preventDefault();
            filterShortcut("currentUser");
          }}
        >
          Created by me
        </button>
        <button
          onClick={(e) => {
            e.preventDefault();
            filterShortcut("myTeam");
          }}
        >
          Created by my team
        </button>
      </div>
      {filterCount > 0 && (
        <Button
          onClick={() =>
            setFilterSearch({
              creator: "",
              category: "" as DibbsConceptType,
              codeSystem: "",
            })
          }
          className={styles.clearFiltersBtn}
          type="button"
        >
          Clear all filters
        </Button>
      )}
    </form>
  );
};

export default DropdownFilter;
