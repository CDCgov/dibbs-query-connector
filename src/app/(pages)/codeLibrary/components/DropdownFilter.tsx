import {
  Dispatch,
  RefObject,
  SetStateAction,
  useEffect,
  useState,
} from "react";
import { Button, Select } from "@trussworks/react-uswds";
import styles from "../codeLibrary.module.scss";
import {
  DibbsConceptType,
  DibbsValueSet,
} from "@/app/models/entities/valuesets";
import {
  formatCodeSystemPrefix,
  formatStringToSentenceCase,
} from "@/app/utils/format-service";
import { emptyFilterSearch } from "../utils";
import { User } from "@/app/models/entities/users";
import {
  getAllGroupMembers,
  getAllUserGroups,
} from "@/app/backend/usergroup-management";
import { applyFocusTrap } from "@/app/ui/utils";

export type FilterCategories = {
  category: DibbsConceptType | undefined;
  codeSystem: string;
  creators: vsAuthorMap;
};

type DropdownFilterProps = {
  setFilterSearch: Dispatch<SetStateAction<FilterCategories>>;
  filterSearch: FilterCategories;
  valueSets: DibbsValueSet[];
  setShowFilters: Dispatch<SetStateAction<boolean>>;
  loading: boolean;
  filterCount: number;
  currentUser: User;
  setTriggerFocus: () => void;
  focusRef: RefObject<HTMLDivElement | null>;
};

export type vsAuthorMap = {
  [groupName: string]: string[];
};
/**
 * @param root0 props
 * @param root0.filterSearch the filter criteria
 * @param root0.setFilterSearch state function to update filter criteria
 * @param root0.setShowFilters state function to toggle showing/hiding the dropdown filters
 * @param root0.valueSets the value sets to apply the filter(s) to
 * @param root0.loading the loading state of the parent's value set data
 * @param root0.filterCount the number of filters currently applied to the result set
 * @param root0.currentUser the User object for the currently active user
 * @param root0.setTriggerFocus function to return focus to the element that triggered the opening of DropdownFilter
 * @param root0.focusRef ref for the DropdownFilter that is accessible to its parent
 * @returns  the DropdownFilter component
 */
const DropdownFilter: React.FC<DropdownFilterProps> = ({
  filterSearch,
  setFilterSearch,
  valueSets,
  setShowFilters,
  loading,
  filterCount,
  currentUser,
  setTriggerFocus,
  focusRef,
}) => {
  const valueSetCodeSystems = valueSets
    .map((vs) => vs.system)
    .filter((item, index, array) => {
      return !!item && array.indexOf(item) === index;
    });

  const [myTeamMembers, setMyTeamMembers] = useState<vsAuthorMap>();

  const [groupAuthors, setGroupAuthors] = useState<vsAuthorMap>({});
  const [valueSetCreators, setValueSetCreators] = useState<vsAuthorMap>({});
  const [focusElements, setFocusElements] = useState<NodeListOf<Element>>();

  useEffect(() => {
    async function mapUsersToGroups() {
      const groups = await getAllUserGroups();
      const authors: vsAuthorMap = {};
      groups.items.map((group) => {
        const members =
          group.members?.map((member) =>
            member.firstName && member.lastName
              ? `${member.firstName} ${member.lastName}`
              : `${member.username}`,
          ) || [];

        authors[group.name] = members;
        return authors;
      });

      setGroupAuthors(authors);
    }

    async function fetchTeammates() {
      const myTeam: {
        [groupName: string]: string[];
      } = {};

      const myGroups = currentUser.userGroupMemberships || [];

      await Promise.all(
        myGroups.map(async (group) => {
          const teammates = await getAllGroupMembers(group.usergroupId);
          const foo = teammates.items.map((user) =>
            user.firstName && user.lastName
              ? `${user.firstName} ${user.lastName}`
              : `${user.username}`,
          );

          myTeam[group.usergroupName] = foo;
          return;
        }),
      );

      setMyTeamMembers(myTeam);
    }

    fetchTeammates();
    mapUsersToGroups();
  }, []);

  useEffect(() => {
    if (groupAuthors) {
      const valueSetAuthors: vsAuthorMap = {};
      valueSets.map((vs) => {
        valueSetAuthors[vs.author] = [vs.author];
      });

      Object.entries(groupAuthors).forEach(([key, val]) => {
        valueSetAuthors[key] = val;
      });

      valueSetAuthors["My Team(s)"] =
        (myTeamMembers && Object.values(myTeamMembers).flat()) || [];
      setValueSetCreators(valueSetAuthors);
    }
  }, [groupAuthors]);

  const getFocusableElements = (focusRef: RefObject<HTMLElement>) => {
    const focusableElements =
      focusRef.current &&
      focusRef.current.querySelectorAll(
        'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])',
      );
    setFocusElements(focusableElements);
  };

  useEffect(() => {
    getFocusableElements(focusRef as RefObject<HTMLElement>);
  }, [filterCount]);

  const valueSetCategories: {
    [dibbsConceptType in DibbsConceptType]: dibbsConceptType;
  } = {
    labs: "labs",
    medications: "medications",
    conditions: "conditions",
  };

  const filterByShortcut = (userIndex: string) => {
    const results = valueSetCreators[userIndex] || [];

    setFilterSearch({
      ...filterSearch,
      creators:
        results.length > 0
          ? { [userIndex]: valueSetCreators[userIndex] }
          : {
              "No creators to filter": ["No creators to filter"],
            },
    });
  };

  const handleOutsideClick = (ref: RefObject<HTMLDivElement | null>) => {
    useEffect(() => {
      function handleClickOutside(event: MouseEvent | KeyboardEvent) {
        if (!ref) return;
        if (
          (event as KeyboardEvent).key == "Escape" ||
          (event.type == "mousedown" &&
            ref.current &&
            !ref.current.contains(event.target as Node))
        ) {
          setTriggerFocus();
          return setShowFilters(false);
        }
      }
      document.addEventListener("mousedown", handleClickOutside);
      document.addEventListener("keyup", handleClickOutside);
      return () => {
        document.removeEventListener("mousedown", handleClickOutside);
        document.removeEventListener("keyup", handleClickOutside);
      };
    }, [ref]);
  };

  handleOutsideClick(focusRef);
  applyFocusTrap(focusRef, focusElements);

  return (
    <div ref={focusRef} className={styles.filtersDropdown}>
      <div className={styles.filterOptions}>
        <div className={styles.filterOptions_grouping}>
          <label htmlFor="category">Category</label>
          <Select
            id="category"
            name="category"
            className={styles.filtersSelect}
            onChange={(e) =>
              setFilterSearch({
                ...filterSearch,
                category: e.target.value as DibbsConceptType,
              })
            }
            value={filterSearch.category}
            disabled={!!loading}
          >
            <option value=""></option>
            {Object.keys(valueSetCategories).map((category) => {
              return (
                <option key={category} value={category}>
                  {formatStringToSentenceCase(category)}
                </option>
              );
            })}
          </Select>
        </div>
        <div className={styles.filterOptions_grouping}>
          <label htmlFor="code-system">Code system</label>
          <Select
            id="code-system"
            className={styles.filtersSelect}
            name="code-system"
            onChange={(e) => {
              e.preventDefault();
              return setFilterSearch({
                ...filterSearch,
                codeSystem: e.target.value as DibbsConceptType,
              });
            }}
            value={filterSearch.codeSystem}
            disabled={!!loading}
          >
            {<option value=""></option>}
            {valueSetCodeSystems.map((codeSystem) => {
              return (
                <option key={codeSystem} value={codeSystem}>
                  {formatCodeSystemPrefix(codeSystem || "")}
                </option>
              );
            })}
          </Select>
        </div>
        <div className={styles.filterOptions_grouping}>
          <label htmlFor="creator">Creator</label>
          <Select
            id="creator"
            name="creator"
            className={styles.filtersSelect}
            onChange={(e) => {
              e.preventDefault();
              valueSetCreators &&
                setFilterSearch({
                  ...filterSearch,
                  creators: {
                    [e.target.value]: valueSetCreators[e.target.value],
                  },
                });
            }}
            value={Object.keys(filterSearch.creators)[0]}
            disabled={!!loading}
          >
            <option value=""></option>
            {valueSetCreators &&
              Object.keys(valueSetCreators).map((key) => {
                return (
                  <option key={key} value={key}>
                    {key}
                  </option>
                );
              })}
          </Select>
        </div>
      </div>
      <div className={styles.shortcuts}>
        <span className="text-italic">Shortcuts:</span>
        <button
          onClick={async (e) => {
            e.preventDefault();
            filterByShortcut(
              currentUser.firstName && currentUser.lastName
                ? `${currentUser.firstName} ${currentUser.lastName}`
                : `${currentUser.username}`,
            );
          }}
        >
          Created by me
        </button>
        <button
          onClick={async (e) => {
            e.preventDefault();

            filterByShortcut("My Team(s)");
          }}
        >
          Created by my team
        </button>
      </div>
      <div className={styles.dropdownButtons}>
        {filterCount > 0 && (
          <Button
            onClick={() => {
              setFilterSearch(emptyFilterSearch);
              if (focusElements?.[0]) {
                (focusElements[0] as HTMLElement).focus();
              }
            }}
            className={styles.clearFiltersBtn}
            type="button"
          >
            Clear all filters
          </Button>
        )}
        <Button
          onClick={() => {
            setShowFilters(false);
            setTriggerFocus();
          }}
          className={styles.clearFiltersBtn}
          type="button"
        >
          Close
        </Button>
      </div>
    </div>
  );
};

export default DropdownFilter;
