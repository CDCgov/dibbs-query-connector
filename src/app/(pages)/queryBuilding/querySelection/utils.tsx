import { ModalRef } from "@/app/ui/designSystem/modal/Modal";
import { DeleteModal } from "@/app/ui/designSystem/modal/deleteModal";
import { JSX, RefObject } from "react";
import { showToastConfirmation } from "@/app/ui/designSystem/toast/Toast";
import { DataContextValue } from "@/app/utils/DataProvider";
import { deleteQueryById } from "@/app/backend/query-building/service";
import { CustomUserQuery } from "@/app/models/entities/query";

/**
 * Handles deleting a user query.
 * @param queryName - The name of the query to delete.
 * @param queryId - The ID of the query to delete.
 * @param queries - The current list of user queries.
 * @param setQueries - Function to update the state of queries.
 * @param context - The data context used to update shared state.
 */
export const handleDelete = async (
  queryName: string | undefined,
  queryId: string | undefined,
  queries: CustomUserQuery[],
  setQueries: React.Dispatch<
    React.SetStateAction<CustomUserQuery[] | undefined>
  >,
  context: DataContextValue | undefined,
) => {
  if (queryId) {
    const result = await deleteQueryById(queryId);
    if (result.success) {
      showToastConfirmation({
        body: `${queryName} has been deleted.`,
        variant: "success",
        duration: 2000,
      });
      const updatedQueries = queries.filter(
        (query) => query.queryId !== queryId,
      );
      setQueries(updatedQueries);

      if (context?.setData) {
        context.setData(updatedQueries);
      }
    } else {
      showToastConfirmation({
        heading: `Something went wrong`,
        body: `${queryName} couldn't be deleted. Please try again or contact us if the error persists`,
        variant: "error",
        duration: 2000,
      });
    }
  } else {
    showToastConfirmation({
      heading: `Something went wrong`,
      body: `${queryName} couldn't be deleted. Please try again or contact us if the error persists`,
      variant: "error",
      duration: 2000,
    });
  }
};

/**
 * Confirms the deletion of a user query by toggling a modal.
 * @param queryName - The name of the query to confirm deletion for.
 * @param queryId - The ID of the query to confirm deletion for.
 * @param setSelectedQuery - Function to set the currently selected query for deletion.
 * @param modalRef - Reference to the modal component.
 */
export const confirmDelete = (
  queryName: string,
  queryId: string,
  setSelectedQuery: React.Dispatch<React.SetStateAction<SelectedQueryDetails>>,
  modalRef: RefObject<ModalRef | null>,
) => {
  setSelectedQuery({ queryName, queryId });
  modalRef.current?.toggleModal();
};

/**
 * Copies the query ID to the clipboard.
 * @param queryName - The name of the query to copy the ID for.
 * @param queryId - The ID of the query to copy.
 */
export const handleCopy = (queryName: string, queryId: string) => {
  navigator.clipboard
    .writeText(queryId)
    .then(() => {
      showToastConfirmation({
        body: `${queryName} ID copied successfully!`,
        duration: 2000,
      });
    })
    .catch((error) => {
      console.error("Failed to copy text:", error);
    });
};

/**
 *  Renders a modal to confirm the deletion of a user query.
 * @param modalRef - Reference to the modal component.
 * @param selectedQuery - The currently selected query for deletion.
 * @param handleDelete - Function to handle the deletion workflow.
 * clean up the internal state after deletion.
 * @param queries - The current list of user queries.
 * @param setQueries - Function to update the state of queries.
 * @param context - The data context used to update shared state.
 * @param setSelectedQuery - Function to update the currently selected query to
 * @returns The JSX element for the modal.
 */
export const renderModal = (
  modalRef: RefObject<ModalRef | null>,
  selectedQuery: SelectedQueryDetails | null,
  handleDelete: (
    queryName: string | undefined,
    queryId: string | undefined,
    queries: CustomUserQuery[],
    setQueries: React.Dispatch<
      React.SetStateAction<CustomUserQuery[] | undefined>
    >,
    context: DataContextValue | undefined,
  ) => void,
  queries: CustomUserQuery[],
  setQueries: React.Dispatch<
    React.SetStateAction<CustomUserQuery[] | undefined>
  >,
  context: DataContextValue,
  setSelectedQuery: React.Dispatch<React.SetStateAction<SelectedQueryDetails>>,
): JSX.Element => {
  // here to fix static build errors when we're not in a browser context
  // if (typeof window === "undefined") {
  //   return <></>;
  // }

  return (
    <DeleteModal
      modalRef={modalRef}
      heading="Confirm Deletion"
      description={`Are you sure you want to delete "${
        selectedQuery ? selectedQuery.queryName : ""
      }"? This action cannot be undone.`}
      onDelete={() => {
        if (selectedQuery) {
          handleDelete(
            selectedQuery.queryName,
            selectedQuery.queryId,
            queries,
            setQueries,
            context,
          );
          setSelectedQuery({ queryName: undefined, queryId: undefined });
        }
      }}
    />
  );
};

export type SelectedQueryDetails = {
  queryName?: string;
  queryId?: string;
  pageMode?: string;
};
export type SelectedQueryState = SelectedQueryDetails | null;
