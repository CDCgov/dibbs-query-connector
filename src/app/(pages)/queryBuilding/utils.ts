import { DibbsValueSet } from "@/app/models/entities/valuesets";
import { ConceptTypeToDibbsVsMap } from "../../utils/valueSetTranslation";

// The structure of the data that's coming from the backend
export type ConditionsMap = {
  [conditionId: string]: {
    name: string;
    category: string;
  };
};

export type CategoryToConditionArrayMap = {
  [category: string]: {
    id: string;
    name: string;
  }[];
};

export type ConditionIdToValueSetArrayMap = {
  [conditionId: string]: DibbsValueSet[];
};

export type NestedQuery = {
  [conditionId: string]: ConceptTypeToDibbsVsMap;
};

export type QueryUpdateResult = {
  id: string;
  query_name: string;
  operation: "INSERT" | "UPDATE";
};
export type QueryDataColumn = {
  [conditionId: string]: { [valueSetId: string]: DibbsValueSet };
};
export type QueryTableResult = {
  queryName: string;
  queryId: string;
  queryData: QueryDataColumn;
  conditionsList: string[];
  immunization: boolean;
};

export const EMPTY_QUERY_SELECTION = {
  queryId: undefined,
  queryName: undefined,
};

export const EMPTY_CONCEPT_TYPE = {
  labs: {},
  conditions: {},
  medications: {},
};

/**
 * Utility method to get a display name for a category (if overridden),
 * or fall back to the raw name.
 * @param categoryName - raw category name from backend
 * @returns display name to show in UI
 */
export function formatCategoryDisplay(categoryName: string): string {
  const categoryDisplayMap: Record<string, string> = {
    "Sexually Transmitted Diseases": "Sexually Transmitted Diseases (STI)",
  };
  // Add more mappings as needed
  return categoryDisplayMap[categoryName] || categoryName;
}

/**
 * Filtering function that checks filtering at the category and the condition level
 * @param filterString - string to filter by
 * @param fetchedConditions - unfiltered list of conditions fetched from the backend
 * @returns - The subset of fetched conditions that contain the filter string
 */
export function filterSearchByCategoryAndCondition(
  filterString: string,
  fetchedConditions: CategoryToConditionArrayMap,
): CategoryToConditionArrayMap {
  const result: CategoryToConditionArrayMap = {};
  const unfilteredConditions = structuredClone(fetchedConditions);

  Object.entries(unfilteredConditions).forEach(
    ([categoryName, conditionArray]) => {
      const displayCategory = formatCategoryDisplay(categoryName);
      const filter = filterString.toLowerCase();

      if (
        categoryName.toLowerCase().includes(filter) ||
        displayCategory.toLowerCase().includes(filter)
      ) {
        result[categoryName] = conditionArray;
      } else {
        const matches = conditionArray.filter((c) =>
          formatDiseaseDisplay(c.name).toLowerCase().includes(filter),
        );
        if (matches.length > 0) {
          result[categoryName] = matches;
        }
      }
    },
  );

  return result;
}

/**
 * Utility method that strips the (disorder) string that comes back from the
 * APHL list on the query building page
 * @param diseaseName - name of the disease
 * @returns A disease display string for display
 */
export function formatDiseaseDisplay(diseaseName: string): string {
  const diseaseDisplayMap: Record<string, string> = {
    "Human immunodeficiency virus infection (disorder)":
      "Human immunodeficiency virus infection (HIV)",
  };
  return (
    diseaseDisplayMap[diseaseName] ||
    diseaseName.replace("(disorder)", "").trim()
  );
}

/**
 * Utility method that returns the number of concepts associated
 * with a given value set, with the option to return only
 * concepts marked as included
 * @param valueSet - the GroupedValueSet to run the tally on
 * @param filterInclude - boolean to indicate whether to only count
 * included concepts (defaults to false)
 * @returns A number indicating the tally of relevant concpets
 */
export function tallyConceptsForSingleValueSet(
  valueSet: DibbsValueSet,
  filterInclude?: boolean,
) {
  const selectedTotal = valueSet.concepts.reduce((sum, concept) => {
    const addToTally = !filterInclude
      ? 1 // add every item
      : concept.include
        ? 1
        : 0; // only add items marked as included

    sum += addToTally;

    return sum;
  }, 0);

  return selectedTotal;
}

/**
 * Utility method that returns the total number of concepts associated
 * with a selection of ValueSets, with the option to return only
 * concepts marked as included
 * @param valueSets - the array of GroupedValueSets to run the tally on
 * @param filterInclude - boolean to indicate whether to only count
 * included concepts
 * @returns A number indicating the tally of relevant concpets
 */
export function tallyConceptsForValueSetArray(
  valueSets: DibbsValueSet[],
  filterInclude?: boolean,
) {
  const selectedTotal = valueSets.reduce((sum, valueSet) => {
    const childTotal = tallyConceptsForSingleValueSet(valueSet, filterInclude);
    sum += childTotal;
    return sum;
  }, 0);
  return selectedTotal;
}

/**
 * Utility method that marks all concepts as included for
 * a given ValueSet
 * @param input - the ValueSet to update
 * @returns the updated ValueSet
 */
export const batchToggleConcepts = (input: DibbsValueSet) => {
  input.concepts.forEach((concept) => {
    const currentStatus = concept.include;
    concept.include = !currentStatus;
  });

  return input;
};
