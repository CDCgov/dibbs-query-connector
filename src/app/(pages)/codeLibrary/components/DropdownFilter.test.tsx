import { render, screen } from "@testing-library/react";
import { RootProviderMock } from "@/app/tests/unit/setup";
import { mockDibbsValueSets } from "../../queryBuilding/fixtures";
import DropdownFilter from "./DropdownFilter";
import {
  mockAdmin,
  mockGroupBasic,
  mockStandard,
  mockSuperAdmin,
} from "../../userManagement/test-utils";
import { emptyVsAuthorMapItem } from "../utils";
import * as UserGroupManagementBackend from "@/app/backend/usergroup-management";
jest.mock("next-auth/react");

jest.mock(
  "@/app/ui/components/withAuth/WithAuth",
  () =>
    ({ children }: React.PropsWithChildren) => <div>{children}</div>,
);

jest.mock("@/app/backend/usergroup-management", () => ({
  getAllUserGroups: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getAllGroupMembers: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
}));

describe("DropdownFilter", () => {
  it("renders correctly", async () => {
    jest
      .spyOn(UserGroupManagementBackend, "getAllGroupMembers")
      .mockResolvedValueOnce({
        items: [mockAdmin, mockStandard, mockSuperAdmin],
        totalItems: 3,
      });
    jest
      .spyOn(UserGroupManagementBackend, "getAllUserGroups")
      .mockResolvedValueOnce({
        items: [mockGroupBasic],
        totalItems: 1,
      });
    render(
      <RootProviderMock currentPage="/codeLibrary">
        <DropdownFilter
          currentUser={mockAdmin}
          filterSearch={{
            category: undefined,
            codeSystem: "",
            creators: emptyVsAuthorMapItem,
          }}
          setFilterSearch={jest.fn()}
          setShowFilters={jest.fn()}
          valueSets={mockDibbsValueSets}
          loading={false}
          filterCount={0}
        />
      </RootProviderMock>,
    );

    const selectDropdowns = await screen.findAllByRole("combobox");
    expect(selectDropdowns[0].id).toBe("category");
    expect(selectDropdowns[1].id).toBe("code-system");
    expect(selectDropdowns[2].id).toBe("creator");

    const buttons = screen.getAllByRole("button");
    expect(buttons).toHaveLength(2);
    expect(buttons[0]).toHaveTextContent("Created by me");
    expect(buttons[1]).toHaveTextContent("Created by my team");
    expect(document.body).toMatchSnapshot();
  });

  it("renders a 'clear filters' button when filters are applied", async () => {
    render(
      <RootProviderMock currentPage="/auditLogs">
        <DropdownFilter
          currentUser={mockAdmin}
          filterSearch={{
            category: "labs",
            codeSystem: "",
            creators: emptyVsAuthorMapItem,
          }}
          setFilterSearch={jest.fn()}
          setShowFilters={jest.fn()}
          valueSets={mockDibbsValueSets}
          loading={false}
          filterCount={1}
        />
      </RootProviderMock>,
    );

    const selectDropdowns = await screen.findAllByRole("combobox");
    expect(selectDropdowns[0].id).toBe("category");
    expect(selectDropdowns[1].id).toBe("code-system");
    expect(selectDropdowns[2].id).toBe("creator");

    const buttons = screen.getAllByRole("button");
    expect(buttons).toHaveLength(3);
    expect(buttons[2]).toHaveTextContent("Clear all filters");

    expect(document.body).toMatchSnapshot();
  });
});
