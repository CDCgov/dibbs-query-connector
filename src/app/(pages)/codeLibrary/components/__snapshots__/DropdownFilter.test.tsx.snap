// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownFilter renders a 'clear filters' button when filters are applied 1`] = `
<body>
  <div>
    <form
      class="filtersDropdown"
    >
      <div
        class="filterOptions"
      >
        <div
          class="filterOptions_grouping"
        >
          <label
            for="category"
          >
            Category
          </label>
          <select
            class="usa-select filtersSelect"
            data-testid="Select"
            id="category"
            name="category"
          >
            <option
              disabled=""
              value=""
            />
            <option
              value="labs"
            >
              Labs
            </option>
            <option
              value="medications"
            >
              Medications
            </option>
            <option
              value="conditions"
            >
              Conditions
            </option>
          </select>
        </div>
        <div
          class="filterOptions_grouping"
        >
          <label
            for="code-system"
          >
            Code system
          </label>
          <select
            class="usa-select filtersSelect"
            data-testid="Select"
            id="code-system"
            name="code-system"
          >
            <option
              disabled=""
              value=""
            />
            <option
              value="http://snomed.info/sct"
            >
              SNOMED
            </option>
          </select>
        </div>
        <div
          class="filterOptions_grouping"
        >
          <label
            for="creator"
          >
            Creator
          </label>
          <select
            class="usa-select filtersSelect"
            data-testid="Select"
            id="creator"
            name="creator"
          >
            <option
              disabled=""
              value=""
            />
            <option
              value="QC Admin"
            >
              QC Admin
            </option>
          </select>
        </div>
      </div>
      <div
        class="shortcuts"
      >
        <span
          class="text-italic"
        >
          Shortcuts:
        </span>
        <button>
          Created by me
        </button>
        <button>
          Created by my team
        </button>
      </div>
      <button
        class="usa-button clearFiltersBtn"
        data-testid="button"
        type="button"
      >
        Clear all filters
      </button>
    </form>
  </div>
</body>
`;

exports[`DropdownFilter renders correctly 1`] = `
<body>
  <div>
    <form
      class="filtersDropdown"
    >
      <div
        class="filterOptions"
      >
        <div
          class="filterOptions_grouping"
        >
          <label
            for="category"
          >
            Category
          </label>
          <select
            class="usa-select filtersSelect"
            data-testid="Select"
            id="category"
            name="category"
          >
            <option
              disabled=""
              value=""
            />
            <option
              value="labs"
            >
              Labs
            </option>
            <option
              value="medications"
            >
              Medications
            </option>
            <option
              value="conditions"
            >
              Conditions
            </option>
          </select>
        </div>
        <div
          class="filterOptions_grouping"
        >
          <label
            for="code-system"
          >
            Code system
          </label>
          <select
            class="usa-select filtersSelect"
            data-testid="Select"
            id="code-system"
            name="code-system"
          >
            <option
              disabled=""
              value=""
            />
            <option
              value="http://snomed.info/sct"
            >
              SNOMED
            </option>
          </select>
        </div>
        <div
          class="filterOptions_grouping"
        >
          <label
            for="creator"
          >
            Creator
          </label>
          <select
            class="usa-select filtersSelect"
            data-testid="Select"
            id="creator"
            name="creator"
          >
            <option
              disabled=""
              value=""
            />
            <option
              value="QC Admin"
            >
              QC Admin
            </option>
            <option
              value="Hogwarts"
            >
              Hogwarts
            </option>
          </select>
        </div>
      </div>
      <div
        class="shortcuts"
      >
        <span
          class="text-italic"
        >
          Shortcuts:
        </span>
        <button>
          Created by me
        </button>
        <button>
          Created by my team
        </button>
      </div>
    </form>
  </div>
</body>
`;
